### ISSUE TITLE
Next.js 14 displays previously cached data if you go back to the page

### ISSUE THREAD

**Issue by Christopher-md** [vercel/next.js#59958]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/next-js-cache-32t3v2?file=%2Fapp%2Fpage.tsx%3A4%2C18

### To Reproduce

1. Open the page where the server request is being executed
2. On the same page, do some actions (Add, Update, etc.)
3. Update the state after step 2
4. Go to another page and go back
5. After go back, Next.js displays the data from step 1

### Current vs. Expected behavior

After the user has returned back, Next.js displays up-to-date data

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: MacOS
  Version: 14.1.2
Binaries:
  Node: 20.1.0
Relevant Packages:
  next: 14.0.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.6
```

### Which area(s) are affected? (Select all that apply)

App Router

### Additional context

I used to use Next.js 13, and `route.refresh()` solved this problem. I found this answer #42546, but it stopped working in Next.js 14

Пробовал использовать [noStore](https://nextjs.org/docs/app/api-reference/functions/unstable_noStore), но это не помогло

**Comment by GregorTB** [comment]
I have the same problem! [Next](https://nextjs.org/), fix it please.

**Comment by JoeyKhd** [comment]
+1

**Comment by zce** [comment]
Due to the current routing cache mechanism, it is like this.

This is the relevant document: https://nextjs.org/docs/app/building-your-application/caching#router-cache

You can follow the discussion of the Next.js team: https://github.com/vercel/next.js/discussions/54075

**Comment by rjsdnql123** [comment]
When the page loads, HTML is dynamically rendered through dynamic rendering, and when navigating routes, Next.js appears to display cached pages, which seems to be functioning correctly.

To resolve the issue, you can either explicitly specify `'use client'` at the top of the file to switch to client-side rendering or use `route.refresh()` to reload the page when needed.

**Comment by Christopher-md** [comment]
> Когда страница загружается, HTML динамически отображается посредством динамического рендеринга, а при навигации по маршрутам Next.js отображает кэшированные страницы, которые, похоже, работают правильно.
> 
> Чтобы решить эту проблему, вы можете либо явно указать `'use client'`в верхней части файла переключение на рендеринг на стороне клиента, либо использовать его `route.refresh()`для перезагрузки страницы при необходимости.

I need SSR, so `use client` is not suitable. Also, `route.refresh()` stopped helping in Next.js 14

**Comment by rjsdnql123** [comment]
@Christopher-md 

I would recommend two alternative solutions:

**Changing the `<Link>` tag in the testPage to `<a>` tag:**

This approach involves requesting new HTML parsing from the server, addressing cache issues. However, it may introduce additional overhead on the server as it requires a fresh request for HTML parsing and JS file retrieval.

**Sending an arbitrary number as a query parameter:**

This method maintains routing between pages while causing the browser to perceive it as a different page, triggering a re-render. It offers a relatively simple solution to resolve cache problems.
ex) `<Link href="/r='random value'">`

When choosing between these two methods, it's advisable to consider the specific circumstances and requirements of your situation to make an informed decision.

**Comment by samcx** [comment]
@Christopher-md As mentioned by @zce, this is due to the Router Cache. I will be closing this in favor of the larger discussion thread! More updates to come to improving this overall, the team has discussed this much internally.
